zerver/webhooks/patreon/view.py

```diff
@@ -0,0 +1,202 @@
+        template = "{user_name} has pledged ${pledge_amount:.2f} per {pay_per_name}. :tada:\nTotal number of patrons: {patron_count}"
+        return template.format(
+            user_name=get_user_name(payload),
+            pledge_amount=pledge_amount,
+            pay_per_name=get_pay_per_name(payload),
+            patron_count=get_patron_count(payload),
+        ).rstrip()
+    raise UnsupportedWebhookEventTypeError(generate_error_msg(event_name, last_charge_status))
+
+
+def get_members_pledge_update_body(payload: WildValue, event_name: str) -> str:
+    last_charge_status = get_last_charge_status(payload)
+    pledge_amount = get_pledge_amount(payload)
+    # The only successful charge status is "Paid". null if not yet charged.
+    if last_charge_status in ("Paid", None) and pledge_amount > 0:
+        template = "{user_name} has updated their pledge to ${pledge_amount:.2f} per {pay_per_name}. :gear:"
+        return template.format(
+            user_name=get_user_name(payload),
+            pledge_amount=pledge_amount,
+            pay_per_name=get_pay_per_name(payload),
+        ).rstrip()
+    raise UnsupportedWebhookEventTypeError(generate_error_msg(event_name, last_charge_status))
+
+
+def get_members_pledge_delete_body(payload: WildValue, event_name: str) -> str:
+    last_charge_status = get_last_charge_status(payload)
+    if last_charge_status in ("Paid", "Deleted", None):
+        template = "{user_name}'s pledge has been cancelled. :cross_mark:\nTotal number of patrons: {patron_count}"
+        return template.format(
+            user_name=get_user_name(payload),
+            patron_count=get_patron_count(payload),
+        ).rstrip()
+    raise UnsupportedWebhookEventTypeError(generate_error_msg(event_name, last_charge_status))
+
+
+def get_last_charge_status(payload: WildValue) -> Optional[str]:
+    return payload["data"]["attributes"]["last_charge_status"].tame(check_none_or(check_string))
+
+
+def get_patron_status(payload: WildValue) -> Optional[str]:
+    return payload["data"]["attributes"]["patron_status"].tame(check_none_or(check_string))
+
+
+def get_user_name(payload: WildValue) -> str:
+    return payload["data"]["attributes"]["full_name"].tame(check_string)
+
+
+def get_pledge_amount(payload: WildValue) -> float:
+    return payload["data"]["attributes"]["currently_entitled_amount_cents"].tame(check_int) / 100
+
+
+def get_patron_count(payload: WildValue) -> int:
+    return payload["included"][0]["attributes"]["patron_count"].tame(check_int)
+
+
+def get_pay_per_name(payload: WildValue) -> str:
+    return payload["included"][0]["attributes"]["pay_per_name"].tame(check_string)
+
+
+def generate_error_msg(
+    event_name: str, last_charge_status: Optional[str], patron_status: Optional[str] = None
+) -> str:  # nocoverage
+    error_msg = f"{event_name}, where last_charge_status={last_charge_status}"
+    if patron_status:
+        error_msg += f", patron_status={patron_status}"
+    return error_msg
```

Comment:
I think that `last_charge_status` is always included as a parameter, so it doesn't need to be optional. And then you can have the if statement just be a return ...

```python
    if patron_status:
        return f"{event_name}, ..."
    return f"{event_name}, ..."
```

Reply:
The thing is that mypy throws an incompatible type error if `last_charge_status` is not `Optional[str]` since `get_last_charge_status` will return either `str` or `None`. But having the if statement just be a return works for me.

Reply:
Right, I wasn't thinking about it being `None` when fetched from the payload. That change works for me.